### ISSUE TITLE
audio OSC support

### ISSUE THREAD

**Issue by scandum** [mintty/mintty#1122]
I'm trying to provide some visually impaired folks with audio, which is rather tedious on WSL.

Would an OSC sequence to play a sound file be something that you're willing to add? Just .wav support would be fine.

Additional features of interest would be volume control and how to handle multiple sounds played at once.

**Comment by mintty** [comment]
Such feature has been proposed in https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/14.
However, no other terminal seems to have implemented anything recently, so the question of how to do it is open.
Also it may be some effort to implement this, but maybe someone would contribute an implementation pattern.

**Comment by scandum** [comment]
That feature describes streaming audio to the terminal.

For my goals an OSC that provides the terminal a filename to play would suffice.

An OSC to play any of the ~30 windows sounds would be great as well. Optionally with a sequence to retrieve a list of available sounds. This seems to be already implemented in mintty through the GUI, so should be relatively easy?

**Comment by jerch** [comment]
Plz note that I see DECPS more as a toy sequencer or programmable bell, that indeed would be fun to have for vintage reasons. But this has nothing in common with "audio OSC support" (I guess OSC here means that audio protocol, not the terminal sequence type).

**Comment by BrianInglis** [comment]
Setting bell frequency and duration and sending \a is a basic capability. 
Didn't Midi devices support serial passthru with some escape codes? 
Some kind of Roland compatible Midi support using portmidi or wildmidi would be a more normal use of serial terminal capabilities. 
Anything more is a terminal feature possibly requiring a heavy infrastructure to support audio sound input and output options.

**Comment by jerch** [comment]
@j4james Indeed, about an audio streaming thingy I cannot find much beside the old RTP protocol and a newer [opus audio format](https://en.wikipedia.org/wiki/Opus_(audio_format)). While the new audio format reads promising in terms of bandwidth / compression ratio it also carries the need for complicated audio processing on both ends (including some patent stuff) - imho thats a no-go for a terminal. A simple PCM sink would avoid that alltogether (ofc for the price of higher bandwidth needs).

**Comment by jerch** [comment]
Ah ok - well I am not yet at the byte encoding front. Ofc a bigger "character room" than base64 would be nice to have. I will have a look at DOCS, maybe we can use that to our advantage. Yet I wonder if that will pass successfully through typical terminal stream layer components (not sure if things like ssh can handle that correctly). Needs investigation.

**Comment by j4james** [comment]
I just found the spec I was looking for, and it's not as simple as I originally thought, but in a good way I think. It actually supports a range of different audio formats and encoding schemes, so you can choose whichever one is most appropriate for your needs. 

The DOCS designation is defined in the ISO-IR register of coded character sets as `ESC % F`. And the audio encoding itself is defined in Appendix E of ITU-T Recommendation T.101 (available [here](https://www.itu.int/rec/T-REC-T.101-199411-I) - the relevant info is in part 3).

>  Yet I wonder if that will pass successfully through typical terminal stream

No, but then I doubt anything will with any degree of certainty. Even with basic OSC sequences, you'll get some terminals that'll just end the sequence early and start writing out the context as text if you continue for too long. And it's not too difficult for an app to test for compatibility with a DOCS encoding by outputting a short stream of blank audio (like a couple of bytes), and then checking if the cursor has moved or not.

**Comment by BrianInglis** [comment]
Any format is useless unless there is already Cygwin/Mingw library support for it, as no terminal developer will spend time building audio codecs or stream processing.

**Comment by jerch** [comment]
@BrianInglis Imho thats not a big issue, as long as we stay away from certain codecs/formats. There are really great platform independent soundlibs for anyone, who does not feel like getting the hands dirty with low level audio handling, like `libsdl` or `PortAudio`. Those libs all have at least PCM8 or PCM16 dual channel interfaces. PortAudio even exposes a blocking IO interface beside the more classic callback driven one. Not sure though about the cygwin lib support (pretty sure sdl is ported, but no clue if the sound hardware is a working condition).

**Comment by BrianInglis** [comment]
The issue with mintty and Cygwin is how much bloat will be dragged in with mintty to add some audio support, as more Windows overhead could affect mintty terminal performance, which is already an issue.

**Comment by jerch** [comment]
Yes, whether thats too much of a burden for mintty is not up to me to decide.

Regarding performance I am more concerned about the windows terminal interfaces like ConPty - not sure if thats fast enough to deliver audio transport at a decent rate (if the data gets through at all, imho there are still issues around DCS/OSC passthrough?)

**Comment by mintty** [comment]
OSC 440; sound BEL would look up a file sound.wav in a mintty sounds resource directory (e.g. ~/.config/mintty/sounds).
Path names are also accepted. OSC 440; sound : async BEL would submit the sound for asynchronous playing etc. OSC 440; BEL stops running asynchronous sound.

**Comment by mintty** [comment]
I've enhanced DECPS support to use the audio output library `libao` if installed, with a number of selectable sound tones.

**Comment by BrianInglis** [comment]
It looks like *Cygwin* library versions need to be installed: `libao4` 1.2.0-3 x86/_64 which pulls in *Pulse Audio* client `libpulse-simple0` and `libpulse0`. 
I expected *mintty* to use the native Windows versions from *Cygwin Mingw* `mingw64-{i686,x86_64}-libao` 1.1.0-1 from `/usr/{i686,x86_64}-w64-mingw32/sys-root/mingw/bin/libao-4.dll`.

**Comment by mintty** [comment]
Hi Brian, thanks for checking this. First, I'd like to point out that I deliberately do *not* pull in any other static dependency. That's why I had to fiddle a lot to make dynamic loading of libao work, via dlopen etc (it uses dlopen itself which spoils the straight-forward approach).
Having said this, I do not see mingw-libao in package category Libs, only in Devel. I do not understand this, can you explain? Do you think it's worth checking that version alternatively to the libao cygwin library? And that many people will install the mingw one but not the other?

**Comment by BrianInglis** [comment]
All *Mingw* library packages are development tools because they require either static linking, or distribution of DLLs with the *Mingw* Windows `exe` using them to where they can be loaded. So maybe they can only be used by *Mingw* builds or *Cygwin* *Mingw* cross-builds and not *Cygwin* builds. I only build them as adjuncts of the packages I maintain, I don't use them, and am unsure what does in *Cygwin*, perhaps this is a service we provide to benefit Mingw developers, and perhaps we should ask this on the apps list? As we will be dropping support for various Windows versions and 32 bit, perhaps we should also drop Mingw builds if they are not part of Cygwin? 
Utilities like `cygcheck`, `setup`, and `strace` use only `msvcrt`, but others: `kill`, `mkgroup`, `mkpasswd`, `passwd`, `ps`, `regtool`, `tzset` and `mintty` use both `msvcrt` and `cygwin1`, although they are *Cygwin* builds. I assume *Cygwin* builds are set up to override `msvcrt` startup if the latter is loaded.
